feat: create database tables and queries for notifications

dannykopping opened the pull request:
Creates two new tables:

- `notification_templates`: templates are used to create notifications from, and specify a unique event in the system
  - right now this will just store immutable system notification templates, but in the future users will be able to configure their own notification templates (e.g. notify when a workspace's disk is 90% full)
- `notification_messages`: each record corresponds to a separate notification which must be delivered to a user
  - this table is effectively a queue, and queue consumers can consume these records concurrently with safety (see `AcquireNotificationMessages`)

`coderd/database/migrations/000219_notifications.up.sql` inserts a single template ("Workspace Deleted"), for which we are creating this vertical slice in this initial stack of PRs.

dannykopping committed:
feat: database tables & queries

dannykopping commented:
* **#13537** <a href="https://app.graphite.dev/github/pr/coder/coder/13537?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#13536** <a href="https://app.graphite.dev/github/pr/coder/coder/13536?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dannykopping and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

mtojek commented on coderd/database/migrations/000219_notifications.up.sql:
```diff
@@ -0,0 +1,66 @@
+(
+    id             uuid                 NOT NULL,
+    name           text                 NOT NULL,
+    enabled        boolean DEFAULT TRUE NOT NULL,
```
Hm.. when exactly can a template be disabled?

dannykopping replied:
This is not necessary for now, actually; this will be used in the future when we allow operators to create their own notifications.

I can remove this if you like.

johnstcn replied:
I'd say remove it for now.

mtojek commented on coderd/database/queries/notifications.sql:
```diff
@@ -0,0 +1,144 @@
+-- name: InsertNotificationTemplate :one
+INSERT INTO notification_templates (id, name, enabled, title_template, body_template, "group")
+VALUES ($1,
+        $2,
+        $3,
+        $4,
+        $5,
+        $6)
```
nit: maybe one liner?

mtojek commented on coderd/database/sqlc.yaml:
```diff
@@ -64,6 +64,12 @@ sql:
+              type: "[]byte"
+          - column: "notification_messages.payload"
+            go_type:
+              type: "[]byte"
```
nit: is it better than string?

dannykopping replied:
These are JSON payloads, and the (un)marshaling steps use `[]byte`.

mtojek commented on coderd/database/migrations/000219_notifications.down.sql:
```diff
@@ -0,0 +1,5 @@
+DROP TABLE IF EXISTS notification_templates;
+DROP TYPE IF EXISTS notification_receiver;
+DROP TYPE IF EXISTS notification_method;
+DROP TYPE IF EXISTS notification_message_status;
```
Can you drop enums safely? I thought that there are some risks around it.

dannykopping replied:
Yup, I tested this again now and it works.

mtojek commented on coderd/database/migrations/000219_notifications.up.sql:
```diff
@@ -0,0 +1,66 @@
+CREATE TYPE notification_message_status AS ENUM (
```
CREATE IF NOT EXISTS?

dannykopping replied:
I'd argue that this _should_ fail if the type already exists.

mtojek replied:
Yeah, we are not consistent there if you look at previous migrations.

mtojek commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -905,6 +905,15 @@ func (*FakeQuerier) AcquireLock(_ context.Context, _ int64) error {
+	if err != nil {
+		return nil, err
+	}
+	// nolint:nilnil // Irrelevant.
```
> 	// nolint:nilnil // Irrelevant.

Is it present somewhere later in the stack?

dannykopping replied:
Yup, in the next PR all of these are implemented; I just stubbed these out so CI passes.

mtojek commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -785,6 +785,13 @@ func (q *querier) AcquireLock(ctx context.Context, id int64) error {
 }
 
+func (q *querier) AcquireNotificationMessages(ctx context.Context, arg database.AcquireNotificationMessagesParams) ([]database.AcquireNotificationMessagesRow, error) {
+	if err := q.authorizeContext(ctx, policy.ActionUpdate, rbac.ResourceSystem); err != nil {
```
cc @johnstcn to validate rbac rules

dannykopping replied:
In the other PR I added a TODO to add a notifications-specific RBAC role; I'll do that as a follow-up.

johnstcn commented on coderd/database/migrations/testdata/fixtures/000219_notifications.up.sql:
```diff
@@ -0,0 +1,23 @@
+-- TODO: add more fixtures
```
I think this is sufficient for now.

johnstcn reviewed (approved):
Some comments below, but nothing blocking.

spikecurtis commented on coderd/database/queries/notifications.sql:
```diff
@@ -0,0 +1,139 @@
+    leased_until     = NULL,
+    next_retry_after = NULL
+FROM (SELECT id, sent_at
+      FROM new_values) AS subquery
```
do we really need both the `WITH` clause and this clause remapping `new_values` to `subquery`?  Is it not supported to just have

```
FROM (SELECT UNNEST(@ids::uuid[])             AS id,
             UNNEST(@sent_ats::timestamptz[]) AS sent_at) as new_values
```

spikecurtis commented on coderd/database/queries/notifications.sql:
```diff
@@ -0,0 +1,139 @@
+       FROM notification_messages AS nested
+       WHERE nested.updated_at < NOW() - INTERVAL '7 days'
+             -- ensure we don't clash with the notifier
+           FOR UPDATE SKIP LOCKED);
```
this `SELECT FOR UPDATE SKIP LOCKED` business is just a performance optimization, and unless we're running these queries very often I don't think it's needed. A plain ole `DELETE` statement will wait for locked rows.  The `SKIP LOCKED` construction causes it to skip locked rows rather than wait for them.

spikecurtis commented on coderd/database/queries/notifications.sql:
```diff
@@ -0,0 +1,139 @@
+-- name: InsertNotificationTemplate :one
```
Is this necessary?  I thought all templates are going to be set by migrations

dannykopping commented on coderd/database/queries/notifications.sql:
```diff
@@ -0,0 +1,144 @@
+        notification_messages
+            SET updated_at = NOW(),
+                status = 'leased'::notification_message_status,
+                status_reason = 'Leased by notifier ' || sqlc.arg('notifier_id')::int,
```
This will be implemented in the PR up the stack.

dannykopping replied:
Done in https://github.com/coder/coder/pull/13536/commits/8a92c8fd92fd41001bd15b700303c45bf7e467de and https://github.com/coder/coder/pull/13537/commits/8b7fa18770343966af00d9b6e101b567ddee430b.

spikecurtis commented on coderd/database/migrations/000219_notifications.up.sql:
```diff
@@ -0,0 +1,66 @@
+    enabled        boolean DEFAULT TRUE NOT NULL,
+    title_template text                 NOT NULL,
+    body_template  text                 NOT NULL,
+    actions        jsonb,
```
Got it now. The upstack PR does use it, I can see. I didn't find it because I was looking at the main `NotificationTemplate` type, and it's used in `FetchNewMessageMetadataRow`

spikecurtis commented on coderd/database/queries/notifications.sql:
```diff
@@ -0,0 +1,144 @@
+                                 SKIP LOCKED
+                         LIMIT sqlc.arg('count'))
+            RETURNING id)
+SELECT
```
One other idea: this query and also the `FetchNewMessageMetadata` query are complicated by the need to pull NotificationTemplate data.  But there is a small, finite number of NotificationTemplates, and we don't allow modifications...  so, we could just lazily load the NotificationTemplates and cache them in the notifier/manager.

That makes the queries simpler to understand (FetchNewMessageMetadata is probably not needed since you can just query the user), and likely faster to execute since Postgres doesn't need to do a join.

At some point down the line if we allow people to modify the templates at runtime, we'd need to invalidate the cache, but that seems pretty simple via the PubSub.

dannykopping replied:
> If that's too strange, I at least think it's clearer to return all the columns of notification_messages we want from the UPDATE, then do a single join to notification_templates, rather than returning the IDs and then doing two joins: back to itself and to notification_templates.

I decided to go for this option; thanks for saving a join 👍 
https://github.com/coder/coder/pull/13536/commits/c1a30103fbaff1d885f1f32e1c6a6459fe286bc0

> One other idea: this query and also the FetchNewMessageMetadata query are complicated by the need to pull NotificationTemplate data. But there is a small, finite number of NotificationTemplates, and we don't allow modifications... so, we could just lazily load the NotificationTemplates and cache them in the notifier/manager.

I agree, but for the sake of simplicity (and inertia, TBH) I think I'll keep this as-is for now and we can optimize/refactor it later.

dannykopping committed:
Change notifier ID to UUID for distributed setups

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Renaming to avoid migration name conflict

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Optimize AcquireNotificationMessages to drop a join

This has the unfortunate side-effect of not allowing the AcquireNotificationMessagesRow.Payload type to be overriden to []byte as it was previously, but it does not change the semantics

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping reviewed (commented):
@spikecurtis I'm back from PTO. I believe I've addressed all of your comments.
Can you please have another look?

spikecurtis commented on coderd/database/queries/notifications.sql:
```diff
@@ -0,0 +1,129 @@
+                           WHEN (attempt_count + 1 < @max_attempts::int)
+                               THEN NOW() + CONCAT(@retry_interval::int, ' seconds')::interval END
+FROM (SELECT id, status, status_reason, failed_at
+      FROM new_values) AS subquery
```
This query is also doing the `WITH new_values` thing where we then essentially rename `new_values` to `subquery`.  Can be simplified to remove the `WITH` clause and unnest the values here in the `FROM` clause.

spikecurtis reviewed (approved):
LGTM

One comment inline, but I don't need to review again.

dannykopping committed:
Simplify BulkMarkNotificationMessagesFailed

Signed-off-by: Danny Kopping <danny@coder.com>
